internal/provider/cluster_resource.go

```diff
@@ -788,66 +804,45 @@ func (r *clusterResource) Update(
 	// CRDB Upgrades/Downgrades
 	if IsKnown(plan.CockroachVersion) && plan.CockroachVersion != state.CockroachVersion {
 		planVersion := plan.CockroachVersion.ValueString()
-		stateVersion := state.CockroachVersion.ValueString()
-		traceAPICall("ListMajorClusterVersions")
-		apiResp, _, err := r.provider.service.ListMajorClusterVersions(ctx, &client.ListMajorClusterVersionsOptions{})
-		if err != nil {
-			resp.Diagnostics.AddError("Couldn't retrieve CockroachDB version list", formatAPIErrorMessage(err))
-			return
-		}
 
-		// Target version must be a valid major version
-		var versionValid bool
-		for _, v := range apiResp.Versions {
-			if v.Version == planVersion {
-				versionValid = true
-				break
-			}
-		}
-		if !versionValid {
-			validVersions := make([]string, len(apiResp.Versions))
-			for i, v := range apiResp.Versions {
-				validVersions[i] = v.Version
+		// Unfortunately the update cluster api endpoint doesn't allow updating
+		// the version and anything else in the same call. Since version may
+		// rely on upgrade_type and plan, if they are updated at the same time
+		// we need to fail out.
+		// We could explore doing the version upgrade after the other updates to
+		// avoid this 2 step apply requirement.
+
+		if client.PlanType(state.Plan.ValueString()) == client.PLANTYPE_BASIC {
+			if client.PlanType(plan.Plan.ValueString()) != client.PLANTYPE_BASIC {
+				resp.Diagnostics.AddError("Error updating cluster", "plan and version must not be changed during the same plan.")
+			} else {
+				resp.Diagnostics.AddError("Error updating cluster", "cockroach_version is not supported for BASIC clusters, please update plan before setting cockroach_version.")
 			}
-			resp.Diagnostics.AddError("Invalid CockroachDB version",
-				fmt.Sprintf(
-					"'%s' is not a valid major CockroachDB version. Valid versions include [%s]",
-					planVersion,
-					strings.Join(validVersions, "|")))
 			return
 		}
 
-		// Next, if we are upgrading, it must be a valid upgrade from the
-		// current version, according to the ListMajorClusterVersions API.
-		if upgrading, err := isUpgrade(stateVersion, planVersion); upgrading {
-			var currentVersionInfo client.ClusterMajorVersion
-			for _, v := range apiResp.Versions {
-				if v.Version == stateVersion {
-					currentVersionInfo = v
-					break
+		if state.ServerlessConfig != nil {
+			stateUpgradeTypeString := state.ServerlessConfig.UpgradeType.ValueString()
+			planUpgradeTypeString := plan.ServerlessConfig.UpgradeType.ValueString()
+			if stateUpgradeTypeString != planUpgradeTypeString {
+				resp.Diagnostics.AddError("Error updating cluster", "upgrade_type and version must not be changed during the same plan.")
+				return
+			} else {
+				stateUpgradeTypePtr, err := client.NewUpgradeTypeTypeFromValue(stateUpgradeTypeString)
+				if err != nil {
+					resp.Diagnostics.AddError("Invalid value for upgrade_type found in state", err.Error())
```

Comment:
looking closer at this, I noticed this is the "error summary" field.  We generally don't add periods for the summary.  But it does seem like this summary is probably too specific.  Here is what would be returned by `err.Error()`:  

```
"invalid value 'BAD_VALUE' for UpgradeTypeType: valid values are ['AUTOMATIC', 'MANUAL']
```

So the error would should up as:

"Invalid value for upgrade_type found in state"
"invalid value 'BAD_VALUE' for UpgradeTypeType: valid values are ['AUTOMATIC', 'MANUAL']

This is repetitive.  I'll update the summary to be "Error updating cluster" like the rest of the errors in this file.   It would be nice to eventually do an audit of the full provider and make a strategy for what to use for the summaries.

Reply:
Great thanks!